File: packages/mirascript/src/vm/types/module.ts
import { kVmModule } from '../../helpers/constants.js';
import { hasOwnEnumerable, keys } from '../../helpers/utils.js';
import type { TypeName, VmAny, VmImmutable } from './index.js';
import { VmWrapper } from './wrapper.js';

/** Mirascript 模块 */
export class VmModule<const T extends Record<string, VmImmutable> = Record<string, VmImmutable>> extends VmWrapper<T> {
    constructor(
        /** 模块名称 */
        readonly name: string,
        /** 模块导出 */
        value: T,
    ) {
        super(value);
    }
    /** @inheritdoc */
    override has(key: string): boolean {
        return hasOwnEnumerable(this.value, key);
    }
    /** @inheritdoc */
    override get(key: string): VmAny {
        if (!this.has(key)) return undefined;
        return this.value[key as keyof T] ?? null;
    }
    /** @inheritdoc */
    override keys(): string[] {
        return keys(this.value);
    }
    /** @inheritdoc */
    override same(other: VmAny): boolean {
        return this === other;
    }
    /** @inheritdoc */
    override get type(): TypeName {
        return 'module';
    }
    /** @inheritdoc */
    override get tag(): string {
        return this.name;
    }
}

Object.defineProperty(VmModule.prototype, kVmModule, { value: true });

File: crates/python/mirascript/vm/types/module.py
# -*- coding: utf-8 -*-
from typing import Any
from .wrapper import VmWrapper

class VmModule(VmWrapper):
	"""
	Mirascript 模块包装器
	"""
	def __init__(self, name: str, value: dict):
		super().__init__(value)
		self.name = name

	def has(self, key):
		return key in self.value

	def get(self, key: str) -> Any:
		if not self.has(key):
			return None
		return self.value.get(key, None)

	def keys(self):
		return list(self.value.keys())

	def same(self, other) -> bool:
		return self is other

	@property
	def type(self):
		return 'module'

	@property
	def describe(self) -> str:
		return self.name
